### ISSUE TITLE
ID Generator

### ISSUE THREAD

**Issue by antonibertel** [vert-x3/issues#197]
Hi,

I would like to have id generator with posibility to persist counter value and restore it when all nodes are died.

**Comment by pmlopes** [comment]
You're looking for a distributed synchronized counter with state. One alternative of having non colliding ids (to some extent, depends on your case) would be something like the `ObjectId` from MongoDB where they do not overlap on a cluster and are ordered you can sort then.

A possible implementation totally independent of 3rd party libraries can be seen here:

https://github.com/pmlopes/vertx-bson-codec/blob/master/src/main/java/com/jetdrone/vertx/codec/bson/ObjectId.java

**Comment by antonibertel** [comment]
So it's good option if you are using mongoDB and you can allow counter to be not integer. 
Thank you.

**Comment by pmlopes** [comment]
The code is independent from mongoDB you can use it for example on a relational DB where the id column is a 12 byte array or 24 byte char array. The interesting thing about it is that you can do natural sorting on it since you can rely on the fact that the id is composed of a timestamp and a sequence (along other info).

**Comment by antonibertel** [comment]
Yes, you are right, in relational DB i can use byte arrays to store key and it's nice that keys are composed and can be sorted. But in any case i will have a small overhead, but it is really minor. (During generation, during search in database, passing to client, etc.)

For me it would be nice to have overloaded method getCounter with possibility to add custom middlewares ->before resultHandler/after resultHandler.
In first case i will have a guaranties that my resultHandler will call only after successful middleware execution. (So i can persist new counter value and make sure that all other instances can use it).
In second case i just want to use new counter ASAP and after possibility to persist using middleware.

But there will be difficulties with restoring persisted value of counter without locks. 
(At the same time ObjectId solved this problem - there is no state, instances don't need synchronization to restore/generate key)
